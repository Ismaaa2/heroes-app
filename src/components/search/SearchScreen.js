import React, { useMemo } from "react";
import { parse } from 'query-string'
import { useLocation } from "react-router";
import { useForm } from "../../hooks/useForm";
import { getHeroesParams } from "../../selectors/getHeroesParams";

export const SearchScreen = ({ history }) => {


  const urlParams = useLocation().search;
  const { q = '' } = parse(urlParams);

  const [valorForm, handleInputChange] = useForm({
    search: q,
  });
  const { search } = valorForm;

  const busquedas = useMemo(() => getHeroesParams(q), [q])


  const handleSubmit = (e) => {
    e.preventDefault();
    history.push( `?q=${search}` );
  };

  return (
    <div>
      <h1>SearchScreen</h1>
      <hr />

      <div className="row">
        <div className="col-5">
          <h4> Search Form </h4>
          <hr />
          <form onSubmit={handleSubmit}>
            <input
              type="text"
              value={search}
              placeholder="Find your hero"
              className="form-control"
              name="search"
              onChange={handleInputChange}
            />
            <button
              type="submit"
              className="btn btn-outline-secondary mt-3 w-100"
            >
              Search...
            </button>
          </form>
        </div>
        <div className="col-7">
          <h4>Results</h4>
          <hr />
          {busquedas.length > 0
            ? busquedas.map((x) => <p className="card-heroe-one" key={x.id}>{x.superhero}</p>)
            : <div className="error-no-data">No hay data</div>}
        </div>
      </div>
    </div>
  );
};
