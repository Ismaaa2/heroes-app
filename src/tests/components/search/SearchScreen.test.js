import { mount } from "enzyme";
import { act } from "react-dom/test-utils";
import { MemoryRouter, Route } from "react-router";
import { SearchScreen } from "../../../components/search/SearchScreen";

describe("<SearchScreen />", () => {


  test("debe de mostrarse correctamente con valores por defecto", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search"]}>
        <Route path="/search" component={SearchScreen} />
      </MemoryRouter>
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(".error-no-data").exists()).toBe(true);
    expect(wrapper.find(".error-no-data").text().trim()).toBe("No hay data");
  });

  test("simular ruta con argumento en url", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search?q=batman"]}>
        <Route path="/search" component={SearchScreen} />
      </MemoryRouter>
    );

    expect(wrapper.find("input").prop("value")).toBe("batman");
    expect(wrapper.find(".card-heroe-one").exists()).toBe(true);
    expect(wrapper).toMatchSnapshot();
  });

  test("debe de mostar un error si no se encuentra el heroe", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search?q=batmanadas"]}>
        <Route path="/search" component={SearchScreen} />
      </MemoryRouter>
    );

    expect(wrapper.find(".error-no-data").exists()).toBe(true);
    expect(wrapper.find(".error-no-data").text().trim()).toBe("No hay data");
    expect(wrapper).toMatchSnapshot();
  });

  test("debe de llamar el push del history", () => {

    const history = {
      push: jest.fn(),
    };

    const wrapper = mount(
      <MemoryRouter initialEntries={["/search"]}>
        <Route
          path="/search"
          component={() => <SearchScreen history={history} />}
        />
      </MemoryRouter>
    );

    wrapper.find('input').simulate('change', {
        target : {
            value: 'green',
            name: 'search'
        }
    });

    act( () => {
        wrapper.find('form').prop('onSubmit')({
            preventDefault(){}
        });
    } )

    expect(history.push).toHaveBeenCalledTimes(1)
    expect(history.push).toHaveBeenCalledWith('?q=green')

  });
});
