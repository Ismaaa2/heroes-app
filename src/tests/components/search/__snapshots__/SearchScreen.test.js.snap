// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchScreen /> debe de mostar un error si no se encuentra el heroe 1`] = `
<div>
  <h1>
    SearchScreen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
         Search Form 
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          className="form-control"
          name="search"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="batmanadas"
        />
        <button
          className="btn btn-outline-secondary mt-3 w-100"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="error-no-data"
      >
        No hay data
      </div>
    </div>
  </div>
</div>
`;

exports[`<SearchScreen /> debe de mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1>
    SearchScreen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
         Search Form 
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          className="form-control"
          name="search"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <button
          className="btn btn-outline-secondary mt-3 w-100"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="error-no-data"
      >
        No hay data
      </div>
    </div>
  </div>
</div>
`;

exports[`<SearchScreen /> simular ruta con argumento en url 1`] = `
<div>
  <h1>
    SearchScreen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
         Search Form 
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          className="form-control"
          name="search"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="batman"
        />
        <button
          className="btn btn-outline-secondary mt-3 w-100"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <p
        className="card-heroe-one"
        key="dc-batman"
      >
        Batman
      </p>
    </div>
  </div>
</div>
`;
